modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,295 @@
+	 *
+	 * @param ILO_URL_Metric $new_url_metric New URL metric.
+	 */
+	public function add( ILO_URL_Metric $new_url_metric ) {
+		$url_metrics = $this->flatten();
+		array_unshift( $url_metrics, $new_url_metric );
+
+		$grouped_url_metrics = $this->group_url_metrics_by_breakpoint( $url_metrics );
+
+		// Make sure there is at most $sample_size number of URL metrics for each breakpoint.
+		$grouped_url_metrics = array_map(
+			function ( $breakpoint_url_metrics ) {
+				if ( count( $breakpoint_url_metrics ) > $this->sample_size ) {
+
+					// Sort URL metrics in descending order by timestamp.
+					usort(
+						$breakpoint_url_metrics,
+						static function ( ILO_URL_Metric $a, ILO_URL_Metric $b ): int {
+							return $b->get_timestamp() <=> $a->get_timestamp();
+						}
+					);
+
+					// Only keep the sample size of the newest URL metrics.
+					$breakpoint_url_metrics = array_slice( $breakpoint_url_metrics, 0, $this->sample_size );
+				}
+				return $breakpoint_url_metrics;
+			},
+			$grouped_url_metrics
+		);
+
+		$this->groups = $grouped_url_metrics;
+	}
+
+	/**
+	 * Groups URL metrics by breakpoint.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics URL metrics.
+	 * @return array<int, ILO_URL_Metric[]> URL metrics grouped by breakpoint. The array keys are the minimum widths for a viewport to lie within
+	 *                                      the breakpoint. The returned array is always one larger than the provided array of breakpoints, since
+	 *                                      the breakpoints reflect the max inclusive boundaries whereas the return value is the groups of page
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] > $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Gets needed minimum viewport widths.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param float $current_time Current time, defaults to `microtime(true)`.
+	 * @return array<int, array{int, bool}> Array of tuples mapping minimum viewport width to whether URL metric(s) are needed.
+	 */
+	public function get_needed_minimum_viewport_widths( float $current_time = null ): array {
```

Comment:
It would be great to find a simpler name for this.

Some ideas:
* `get_incomplete_minimum_viewport_widths()`
* `get_incomplete_viewport_widths()`
* `get_incomplete_group_identifiers()` (which should be clear enough given that the groups are always by minimum viewport width)

Reply:
I think `needed` is fine because it's relevant not only if the group doesn't have the sample size but also if any of the URL metrics in the group are stale.

Reply:
I spent some time thinking about this some more. I just pushed up some more changes. Now instead of `get_needed_minimum_viewport_widths()` the method is called `get_viewport_group_lacking_statuses()`. Not any less verbose. But maybe more clear?